### PR TITLE
Use single file write when an extension is present in the path.

### LINKED ISSUE #9684: `to_parquet` with path not ending in a slash writes to a file instead of a directory since v36
### Describe the bug

`Dataframe::to_parquet` takes a path as string, and used to create a directory and write files to it, even if it did not end with `/`. Starting with v36, it writes to a single file if the path does not end with `/`.

### To Reproduce

(using the Python library)

```python
from pathlib import Path
import shutil

import datafusion
import pyarrow.dataset

target_path = Path("/tmp/") / "dataset"

# clean up any existing data at /tmp/dataset
if target_path.is_file():
    target_path.unlink()
elif target_path.is_dir():
    shutil.rmtree(target_path)
assert not target_path.exists()

ctx = datafusion.SessionContext()

ctx.sql("SELECT 1").write_parquet(str(target_path))

assert target_path.exists()

if target_path.is_file():
    print(f"{target_path} is a file")
else:
    print(
        f"{target_path} is a directory with entries:",
        ", ".join(map(str, target_path.iterdir()))
    )

# clean up /tmp/dataset
if target_path.is_file():
    target_path.unlink()
elif target_path.is_dir():
    shutil.rmtree(target_path)
assert not target_path.exists()
```

v36 behavior:

```
/tmp/dataset is a file
```

### Expected behavior

v35 behavior:

```
/tmp/dataset is a directory with entries: /tmp/dataset/u6mIvvxwwFbjo17Q_0.parquet
```

### Additional context

It is common to use the [`pathlib`](https://docs.python.org/3/library/pathlib.html) module in Python when working with paths, and it always strips the trailing `/`.

### PR DESCRIPTION
**Pull request by dhegberg**
## Which issue does this PR close?

Closes #9684.

## Rationale for this change

Dataframe's `write_parquet()` was identified as incorrectly identifying paths without an extension as a single file output.

This change updates `start_demuxer_task` to respect the suggested behaviour:

```
    tmp/dataset/ -> is a folder since it ends in /
    tmp/dataset -> is still a folder since it does not end in / but has no valid file extension
    tmp/file.parquet -> is a file since it does not end in / and has a valid file extension .parquet
    tmp/file.parquet/ -> is a folder since it ends in /

```

## What changes are included in this PR?

* Add `file_extension()` to `ListingTableUrl` to return an Optional extension
* Update `start_demuxer_task()` to require the presence of an extension from the `ListingTableUrl` to set `single_file_output` to true
* Rename `file_extension` to `default_extension` to indicate usage will be ignored if `single_file_output` is triggered.

## Are these changes tested?
* Unit tests added for `file_extension()`
* Unit tests added for `Dataframe.write_parquet()` for paths with and without extensions.
* No direct testing for `start_demuxer_task` since there was no direct testing originally.  I can revise and test this directly if preferred.

Testing via `cargo test -- --test-threads=1`

## Are there any user-facing changes?
* Yes, the file output write behaviour is changing.

### CODE REVIEW

**Comment by alamb** [comment]
Thanks @dhegberg  -- I plan to review this later today

File: datafusion/core/src/datasource/listing/url.rs [hunk 1]
```diff
@@ -493,4 +506,54 @@ mod tests {
             "path not ends with / - fragment ends with / - not collection",
         );
     }
+
+    #[test]
+    fn test_file_extension() {
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/url.rs:511]: Really nice tests 👏

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 2]
```diff
@@ -2338,6 +2298,140 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/parquet.rs:2301]: This is really nice too

**Review by alamb — APPROVED** [review]
Thank you @dhegberg  -- this is a really nice PR -- I think the code and tests are well written.

Thank you 🙏 

cc @progval 


I also tried it locally with datafusion-cli and it works as expected 👌 

```shell
> copy (values (1), (2)) to '/tmp/foo' STORED AS parquet;
+-------+
| count |
+-------+
| 2     |
+-------+
1 row(s) fetched.
Elapsed 0.030 seconds.

>
\q
(venv) andrewlamb@Andrews-MacBook-Pro-2:~/Software/datafusion/datafusion-cli$ ls -ltr /tmp/foo
total 8
-rw-r--r--@ 1 andrewlamb  wheel   342B Nov  1 12:31 MrzgxU8HT1fn3wTB_0.parquet
(venv) andrewlamb@Andrews-MacBook-Pro-2:~/Software/datafusion/datafusion-cli$
```

**Comment by alamb** [comment]
Here is a small follow on PR to add some more docs https://github.com/apache/datafusion/pull/13216 (really get the great writeup you did on this PR into the code0

**Comment by sergiimk** [comment]
I suspect this introduced a regression - would appreciate your opinion on #13323

**Pull request merged**